clippy_lints/src/vec.rs

```diff
@@ -110,31 +97,62 @@ impl<'tcx> LateLintPass<'tcx> for UselessVec {
             .is_continue();
 
             if only_slice_uses {
-                self.check_vec_macro(
-                    cx,
-                    &vec_args,
-                    expr.span.ctxt().outer_expn_data().call_site,
-                    SuggestedType::Array,
-                );
+                let span = expr.span.ctxt().outer_expn_data().call_site;
+                self.lintable_exprs.insert(expr.hir_id);
+                self.check_vec_macro(cx, &vec_args, span, expr.hir_id, SuggestedType::Array);
             }
         }
-
         // search for `for _ in vec![…]`
-        if let Some(higher::ForLoop { arg, .. }) = higher::ForLoop::hir(expr)
+        else if let Some(higher::ForLoop { arg, .. }) = higher::ForLoop::hir(expr)
             && let Some(vec_args) = higher::VecArgs::hir(cx, arg)
             && self.msrv.meets(msrvs::ARRAY_INTO_ITERATOR)
         {
             // report the error around the `vec!` not inside `<std macros>:`
             let span = arg.span.ctxt().outer_expn_data().call_site;
-            self.check_vec_macro(cx, &vec_args, span, SuggestedType::Array);
+            self.lintable_exprs.insert(arg.hir_id);
+            self.check_vec_macro(cx, &vec_args, span, arg.hir_id, SuggestedType::Array);
+        }
+        // search for `&vec![_]` or `vec![_]` expressions where the adjusted type is `&[_]`
+        else if let Some(vec_args) = higher::VecArgs::hir(cx, expr.peel_borrows()) {
```

Comment:
The first check of this method was specifically looking for cases of `let x = vec![...]` where every subsequent use of `x` adjusts to a slice, whereas this check looks for cases where `vec![]` is used in a context where the expression's type adjusts to a slice.

The original lint was also set up in this manner, but I readjusted the order of the different checks and also changed subsequent checks to use "else if" instead of "if".

Reply:
But I don't understand why you chose to create a new branch checking `if let Some(vec_args) = higher::VecArgs::hir(cx, expr.peel_borrows())`, instead of adding this logic to the previous that checks the same thing.